src/lib.rs

```diff
@@ -187,3 +187,71 @@ pub enum ReflinkSupport {
     /// Reflink support is unconfirmed.
     Unknown,
 }
+
+/// Creates a reflink of a specified block from one file to another.
+///
+/// This function is designed to be high performant and does not perform any extra API calls.
+/// It is expected that the user takes care of necessary preliminary checks and preparations.
+///
+/// If you need to clone an entire file, consider using the [`reflink`] or [`reflink_or_copy`]
+/// functions instead.
+///
+/// > Note: Currently the function works only for windows. It returns `Err` for any other platform.
+///
+/// # Windows Restrictions and Remarks
+/// - The source and destination regions must begin and end at a cluster boundary.
+/// - The cloned region must be less than 4GB in length.
+/// - The destination region must not extend past the end of file. If the application wishes to
+///   extend the destination with cloned data, it must first call
+///   [`File::set_len`](fn@std::fs::File::set_len).
+/// - If the source and destination regions are in the same file, they must not overlap. (The
+///   application may able to proceed by splitting up the block clone operation into multiple block
+///   clones that no longer overlap.)
+/// - The source and destination files must be on the same ReFS volume.
+/// - The source and destination files must have the same Integrity Streams setting (that is,
+///   Integrity Streams must be enabled in both files, or disabled in both files).
+/// - If the source file is sparse, the destination file must also be sparse.
+/// - The block clone operation will break Shared Opportunistic Locks (also known as Level 2
+///   Opportunistic Locks).
+/// - The ReFS volume must have been formatted with Windows Server 2016, and if Windows Failover
+///   Clustering is in use, the Clustering Functional Level must have been Windows Server 2016 or
+///   later at format time.
+///
+/// More information can be found by the
+/// [link](https://learn.microsoft.com/en-us/windows/win32/fileio/block-cloning).
+///
+/// # Examples
+///
+/// ```no_run
+/// use std::fs::File;
+///
+/// fn main() -> std::io::Result<()> {
+///     const CLUSTER_SIZE: u64 = 4096;
+///     let from_file = File::open("source.txt")?;
+///     let len = from_file.metadata()?.len();
+///     let to_file = File::create("destination.txt")?;
+///     to_file.set_len(len)?;
+///     let mut offset = 0u64;
+///     while offset < len {
+///         reflink_copy::reflink_block(&from_file, offset, &to_file, offset, CLUSTER_SIZE)?;
+///         offset += CLUSTER_SIZE;
+///     }
+///     if offset > len {
+///         to_file.set_len(len)?;
+///     }
+///     Ok(())
+/// }
+/// ```
+#[cfg_attr(not(windows), allow(unused_variables))]
+pub fn reflink_block(
+    from: &fs::File,
+    from_offset: u64,
+    to: &fs::File,
+    to_offset: u64,
+    block_size: u64,
+) -> io::Result<()> {
```

Comment:
I kind of feel like this API is too specific to windows.

I checked the linux API https://man7.org/linux/man-pages/man2/ioctl_ficlonerange.2.html and there's no `block_size`

Feel like we should have use builder pattern to abstract it, with a windows-only function to specify `block_size` with a default value?

Reply:
> This ioctl reflinks up to src_length bytes from file descriptor src_fd at offset src_offset into the file dest_fd at offset dest_offset, provided that both are files.
> If src_length is zero, the ioctl reflinks to the end of the source file.


`src_length` specifies how many bytes to copy over, not the block size.

Reply:
The intention of this function is to reflink a portion of a file to another location which I think makes sense for both platforms. Maybe there is a problem with the naming, instead of block_size, it might be more appropriate to use src_length.

Overall, I believe this function can work similarly on both Windows and Linux with some additional restrictions. One challenge I've encountered is the unknown filesystem cluster size. Making an additional API call for each invocation seems too costly. Passing the cluster size as an argument would allow for prerequisite checks and handling more cases efficiently.

Proposed changes:
- Rename block_size to src_length
- Return an Err if src_length is zero
- Pass the cluster size as an additional argument
- Add a loop to handle lengths greater than 4GB

Reply:
> Maybe there is a problem with the naming, instead of block_size, it might be more appropriate to use src_length.

Ah so that's actually the number of bytes to copy?

> Passing the cluster size as an argument would allow for prerequisite checks and handling more cases efficiently.

Maybe we can have a builder-pattern?

Cluster-size can be windows-only function, if left unspecified then it would be automatically detected.

> - Return an Err if src_length is zero

I think we can keep this behavior while remain zero-cost, by using builder pattern to enable optional parameters.

On windows, if left unspecified, an additional syscall would be made to get the length.

But if simply banning it is easier, we could go with it for now, and use `NonZeroU64`, so that you don't even have to handle this scenario.